priv/prelude/erlang.specs.erl

```diff
@@ -15,3 +15,33 @@
 
 %% -spec tl([term(), ...]) -> [term()].
 -spec tl([A, ...]) -> [A].
+
+-spec erlang:'--'(list(), list()) -> list().
+
+%% The original spec is:
+%%
+%% -spec erlang:'++'(list(), term()) -> term().
+%%
+%% Now, this is funny:
+%%
+%%   > [] ++ b.
+%%   b
+%%   > [a] ++ b.
+%%   [a|b]
+%%   > [a, b] ++ c.
+%%   [a,b|c]
+%%   > [a|b] ++ c.
+%%   ** exception error: bad argument
+%%     in operator  ++/2
+%%        called as [a|b] ++ c
+%%   > [a, b] ++ [c].
+%%   [a,b,c]
+%%   > [a|b] ++ [c].
+%%   ** exception error: bad argument
+%%        in operator  ++/2
+%%           called as [a|b] ++ [c]
+%%
+-spec erlang:'++'(list(T1), list(T2)) -> list(T1 | T2);
+                 (list(T1), nonempty_improper_list(T2, T3)) -> nonempty_improper_list(T1 | T2, T3);
+                 ([], T) -> T;
+                 (nonempty_list(T1), T2) -> nonempty_improper_list(T1, T2).
```

Comment:
According to docs, the clauses in a spec are not allowed to overlap. It's not like a case expression where the first matching clause is chosen.

I don't know if we can write such spec though. Perhaps we should go for this one.

Would you care to add some examples (test cases) to show that this fancy spec is useful?

Reply:
> Would you care to add some examples (test cases) to show that this fancy spec is useful?

Hmm, I tried coming up with some, but apparently this spec is not taken into account at all. [There are hardcoded rules for handling `++` and `--`](https://github.com/josefs/Gradualizer/blob/master/src/typechecker.erl#L2940-L2947). So for the sake of #437 the only important thing is to have any spec defined. I'll work on some known problems tests for this.

Reply:
Yeah I thought there are special code for binops. Perhaps the spec ia needed only when called as a funcrion ie like `erlang:'++'(X, Y)`...?

Reply:
Ok, this seems to be a bigger topic previously being discussed in #163 (closed today by Pierre). I think in this PR we should focus on unblocking #437 and maybe providing known_problems for some list operators, but in general not let it linger for too long.

Reply:
Ok, I have added the tests following the examples and the sketched out spec. Indeed, the function forms seem to already be checked against the spec, but not the operator forms.

Reply:
Well, officially Erlang's type/spec language is specific to Dialyzer. We just happen to reuse that.

According to some paper I've read (the one about gradual typing with union, intersection and negation types IIRC), it is problematic to have overlapping domains. (I don't remember exactly why, but there is a way to rewrite the code to exclude the domains of the previous clauses using intersection and negation types.)

Regarding custom code vs specs for operators, it was decided that it's useful to have some richer semantics for some operators than what the type language can express. I don't have a strong opinion but think we can allow custom code for operators and for functions too, for some bifs. The distinction between binary operators and function calls is just syntactic.

Reply:
> it's useful to have some richer semantics for some operators

So far it seems we're seeing incomplete implementations instead of richer semantics ;p As in the case of `++` whose type checking implementation is less accurate than the spec I sketched above. Anyway, I think that we can have custom implementations for "some operators" where it makes sense yet default to treating the operators and `erlang:*` invocations as equivalent.

> According to some paper [...]

Yeah, I remember reading that, too. That's a convincing point, indeed. If they overlap it's just not clear (not possible to tell deterministically) which spec to choose.

Ultimately, it's possible to rewrite the spec not to have overlapping domains. It's going to have more clauses, so will be slightly less readable, but it should still match the new tests.

Reply:
> Ultimately, it's possible to rewrite the spec not to have overlapping domains.

Ok, I take it back. It's not possible without negation types. `T2` in the third clause overlaps the 2nd param of the first and second clauses...

```
-spec erlang:'++'([T1, ...], [T2])                           -> [T1 | T2, ...];
                 ([T1, ...], nonempty_improper_list(T2, T3)) -> nonempty_improper_list(T1 | T2, T3);
                 ([T1, ...], T2)                             -> nonempty_improper_list(T1, T2);
                 ([], T)                                     -> T.
```

Reply:
The thing I wanted to achieve with the spec was the type of a proper `[a,b,c]` ++ `[d,e,f]` to be a proper list. Due to the implementation of the `++` operator and improper lists it turns out we cannot escape the `_ ++ _ :: term()` :/

Reply:
I know. I believe Josef would just say that ++ should be used for proper lists only. Anything else is a type error. (But I'm not sure it's correct given you actually use it for other stuff even if we don't like it.)

Reply:
We can go for the spec you came up with. We can always change it later if it poses any problems.